modules/core/perf/perf_math.cpp

```diff
@@ -36,6 +36,227 @@ PERF_TEST_P(VectorLength, phase64f, testing::Values(128, 1000, 128*1024, 512*102
     SANITY_CHECK(angle, 5e-5);
 }
 
+// generates random vectors, performs Gram-Schmidt orthogonalization on them
+Mat randomOrtho(int rows, int ftype, RNG& rng)
+{
+    Mat result(rows, rows, ftype);
+    rng.fill(result, RNG::UNIFORM, cv::Scalar(-1), cv::Scalar(1));
+
+    for (int i = 0; i < rows; i++)
+    {
+        Mat v = result.row(i);
+
+        for (int j = 0; j < i; j++)
+        {
+            Mat p = result.row(j);
+            v -= p.dot(v) * p;
+        }
+
+        v = v * (1. / cv::norm(v));
+    }
+
+    return result;
+}
+
+template<typename FType>
+Mat buildRandomMat(int rows, int cols, RNG& rng, int rank)
+{
+    int mtype = cv::traits::Depth<FType>::value;
+    Mat u = randomOrtho(rows, mtype, rng);
+    Mat v = randomOrtho(cols, mtype, rng);
+    Mat s(rows, cols, mtype, Scalar(0));
+
+    std::vector<FType> singVals(rank);
+    rng.fill(singVals, RNG::UNIFORM, Scalar(0), Scalar(10));
+    std::set<FType> setSingVals(singVals.begin(), singVals.end());
+    auto singIter = setSingVals.rbegin();
+    for (int i = 0; i < rank; i++)
+    {
+        s.at<FType>(i, i) = *singIter++;
+    }
```

Comment:
cv::Mat::diag does the same:  https://docs.opencv.org/4.x/d3/d63/classcv_1_1Mat.html#ae22383dd8cc054d6f47a672821309391

Reply:
`Mat::diag()` is similar but that flexible. I want singular values to be sorted in descending order and make the code independent of their actual amount (which is not always equal to matrix size) which is not what `Mat::diag()` does.

Reply:
In case of duplication in singVals (it's possible with rng) we get out-of-bound access in std::set or exception.
I propose simpler solution:
```
std::vector<FType> singVals(rank);
    rng.fill(singVals, RNG::UNIFORM, Scalar(0), Scalar(10));
std::sort(singVals.begin(), singVals.end());
std::Mat s = cv::Mat::diag(singVals);
```

Reply:
Sure, segfault has to be fixed
But using `Mat::diag()` is still not an option because:
* s matrix is not necessarily square
* `rank` can be less than any of matrix dimensions